### ISSUE TITLE
Tests fail in libwacom-0.33 due to duplicate Huion devices

### ISSUE THREAD

**Issue by awilfox** [linuxwacom/libwacom#85]
```
FAIL: test-load
===============


(process:25502): libwacom-CRITICAL **: 21:06:57.024: Duplicate match of 'usb:256c:006e' on device 'Huion H420'.
Assertion failed: db (test-load.c: main: 77)
Failed to load data from /usr/src/packages/user/libwacom/src/libwacom-0.33/dataFAIL test-load (exit status: 134)

FAIL: test-dbverify
===================


(process:25503): libwacom-CRITICAL **: 21:06:57.025: Duplicate match of 'usb:256c:006e' on device 'Huion H420'.
Assertion failed: db (test-dbverify.c: main: 194)
Failed to load data from /usr/src/packages/user/libwacom/src/libwacom-0.33/dataFAIL test-dbverify (exit status: 134)

FAIL: test-tablet-validity
==========================


(process:25504): libwacom-CRITICAL **: 21:06:57.026: Duplicate match of 'usb:256c:006e' on device 'Huion H420'.
Assertion failed: db (test-tablet-validity.c: main: 270)
Failed to load data from /usr/src/packages/user/libwacom/src/libwacom-0.33/dataFAIL test-tablet-validity (exit status: 134)
```

This seems to be caused by:

```
src/libwacom-0.33/data/huion-h610-pro.tablet
7:DeviceMatch=usb:256c:006e:HUION PenTablet Pen;usb:256c:006e:HUION PenTablet Pad;usb:256c:006e:Turcom TS-6610 Pen;usb:256c:006e:Turcom TS-6610 Pad

src/libwacom-0.33/data/huion-h420.tablet
17:DeviceMatch=usb:256c:006e:HUION H420 Pen;usb:256c:006e:HUION H420 Pad;
```

Removing the huion-h420.tablet and layout/SVG files allows libwacom to pass its test suite and function again.

**Comment by jigpu** [comment]
I don't seem to be able to reproduce this issue. I tried running `make check` from both a fresh checkout of libwacom-0.33 from git as well as the libwacom-0.33 release tarball.

**Comment by pickfire** [comment]
I seemed to have 3 the same tests failed when packaging version 0.33 for alpine linux.

**Comment by whot** [comment]
The relevant bit of code here is:
```
                      matchstr = libwacom_match_get_match_string(*match);                                              
                      /* no duplicate matches allowed */                         
                      if (g_hash_table_lookup(db->device_ht, matchstr) != NULL) {  
                              g_critical("Duplicate match of '%s' on device '%s'.",                               
                                          matchstr, libwacom_get_name(d));
                              libwacom_database_destroy(db);
```

And running this with a `printf` here shows the match strings to include the name:
```
$ ./test/test-dbverify | grep 006e
... matchstr: usb:256c:006e:HUION PenTablet Pen
... matchstr: usb:256c:006e:HUION PenTablet Pad
... matchstr: usb:256c:006e:Turcom TS-6610 Pen
... matchstr: usb:256c:006e:Turcom TS-6610 Pad
... matchstr: usb:256c:006e:HUION H420 Pen
... matchstr: usb:256c:006e:HUION H420 Pad
... matchstr: usb:256c:006e:HUION PenTablet Pen
... matchstr: usb:256c:006e:HUION PenTablet Pad
... matchstr: usb:256c:006e:Turcom TS-6610 Pen
... matchstr: usb:256c:006e:Turcom TS-6610 Pad
... matchstr: usb:256c:006e:HUION H420 Pen
... matchstr: usb:256c:006e:HUION H420 Pad
```

In your output only the bits without the name are printed as `matchstr`, this suggests that `match_from_string()` fails to parse them correctly. Would be best to gdb this and figure out what's going on there, because there's nothing obvious here that should fail.

**Comment by awilfox** [comment]
Oh, yes, I see now.  This is definitely a bug in libwacom.  Same one as [git](https://www.openwall.com/lists/musl/2017/09/16/1); see that thread for the C99 specification that is being broken here.

Unfortunately, the fix won't be so simple; replacing 'c' with 's' will stop at the first whitespace character.

**Comment by awilfox** [comment]
An awful, cheesy hack is to replace "%63c" with "%63[^\b]" or some other control character.

**Comment by whot** [comment]
oh, nice find, thank you.

TLDR: `scanf("%63c")` means "read up to 63 characters" in *most* implementations, but musl reads it as **exactly** 63 characters and fails if it's less than that. musl derives this from the POSIX standard and is technically correct here.

best solution is probably to tokenize the string by the first three `:` and then handle each token separately. Want to submit a patch for that? :)

**Comment by jigpu** [comment]
Thought I had a nice solution based around `%n` but then a quirk of the C standard had to rear its head:
~~~patch
diff --git a/libwacom/libwacom-database.c b/libwacom/libwacom-database.c
index afb89a2..28388f5 100644
--- a/libwacom/libwacom-database.c
+++ b/libwacom/libwacom-database.c
@@ -147,14 +147,13 @@ make_match_string (const char *name, WacomBusType bus, int vendor_id, int produc
 static gboolean
 match_from_string(const char *str, WacomBusType *bus, int *vendor_id, int *product_id, char **name)
 {
-	int rc = 1;
-	char busstr[64], namestr[64];
+	int rc = 1, len = 0;
+	char busstr[64];
 
-	memset(namestr, 0, sizeof(namestr));
-
-	rc = sscanf(str, "%63[^:]:%x:%x:%63c", busstr, vendor_id, product_id, namestr);
-	if (rc == 4) {
-		*name = g_strdup(namestr);
+	rc = sscanf(str, "%63[^:]:%x:%x:%n", busstr, vendor_id, product_id, &len);
+	if (len > 0) {
+		/* Grumble grumble scanf handling of %n */
+		*name = g_strdup(str+len);
 	} else if (rc == 3) {
 		*name = NULL;
 	} else {
~~~
If it passes your smell test though, I can submit it properly.

**Comment by whot** [comment]
LGTM, thanks.

@awilfox, any comments?

**Comment by awilfox** [comment]
This looks right to me, and does fix the issue.  Thanks!

**Comment by jigpu** [comment]
Fix merged

**Comment by 4a454646** [comment]
**Note for any future users:**
It doesn't seem to be fixed on 2.1.0, and this error seems to prevent gnome-control-center (GNOME settings) from starting up completely!
However the very simple fix that worked for me was to simply unplug the tablet, then start Settings.
